_blog/2024-01-18-parsing-json-rust.mdx

```diff
@@ -0,0 +1,229 @@
+    timestamp_last_updated: Datetime<Utc>,
+	  categories: Vec<String>,
+}
+```
+
+The most common use case for this would be nested JSON in a web service. For example, when you are receiving a POST request to your API that has a JSON body, you would normally pass the relevant `Json` type in as a handler function parameter. See below:
+
+```rust
+use axum::Json;
+use chrono::{DateTime, Utc};
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize)]
+pub struct Post {
+	  nested_json: PostMetadata,
+    title: String,
+	  body: String
+}
+
+#[derive(Serialize, Deserialize)]
+pub struct PostMetadata {
+	  timestamp_created: DateTime<Utc>,
+    timestamp_last_updated: Datetime<Utc>,
+	  categories: Vec<String>,
+}
+
+async fn receive_some_json(
+  // this extractor consumes a JSON body and converts it into the struct type given
+	Json(json): Json<Post>
+) -> Json<Post> {
+  println!("{:?}", json);
+	Json(json)
+}
+```
+
+In addition to the previous code snippet that shows how you can use `serde_json` to convert to a struct from a JSON string, you can also convert to a struct from its byte representation:
+
+```rust
+let json_as_bytes = b"
+        {
+            \"message\": \"Hello world!\",
+        }";
+
+    let my_struct: MyStruct = serde_json::from_slice(json_as_bytes).unwrap();
+```
+
+This is particularly useful if you want to store a struct somewhere as a byte array and then turn it back into a struct later on!
+
+Similarly, you can also read JSON and turn it into a struct from a IO stream of JSON using the `.from_reader()` method. Below is an example taken from the `serde_json` docs for how you could use it with a TCP stream:
+
+```rust
+use serde::Deserialize;
+use std::error::Error;
+use std::net::{TcpListener, TcpStream};
+
+#[derive(Deserialize, Debug)]
+struct User {
+    fingerprint: String,
+    location: String,
+}
+
+fn read_user_from_stream(tcp_stream: TcpStream) -> Result<User, Box<dyn Error>> {
+    let mut to_be_deserialized = serde_json::Deserializer::from_reader(tcp_stream);
+    let user = User::deserialize(&mut to_be_deserialized)?;
+
+    Ok(user)
+}
+
+fn main() {
+    let listener = TcpListener::bind("127.0.0.1:4000").unwrap();
+
+    for stream in listener.incoming() {
+        println!("{:#?}", read_user_from_stream(stream.unwrap()));
+    }
+}
+```
+
+Doing it this way allows you to deserialize from the stream directly instead of adding buffering in memory. If you’re receiving a lot of JSON-based data, this can help you quite a bit!
+
+## Comparing Rust JSON crates
+
+Although `serde-json` may be the most popular crate, it is by no means the fastest. A few other crates have popped up in the meantime to improve general JSON parsing performance. In exchange for performance, however, there are certain caveats regarding CPU architecture requirements. There is also increased use of unsafe code, though generally speaking best efforts have been upheld to make sure the code is safe to use. 
```

Comment:
What do you mean by `certain caveats regarding CPU architecture requirements`? Just wondering if we could be more specific somehow. E.g. if the requirements are related to libraries that are compatible only with certain CPU architectures, maybe we can say something like: `there are certain caveats regarding which platforms the libraries can be built & used on`.

LE: I see now what you mean by CPU architecture requirements (e.g. non-parallelized CPU usage). In this case, I would phrase it as: `there are certain caveats regarding CPU SIMD extension requirements`.